### PR TITLE
[mono] Add iOS app sample using mono runtime

### PR DESCRIPTION
**Pull request by mdh1418**
## Summary

This PR looks to port over the iOS [sample](https://github.com/dotnet/runtime/tree/master/src/mono/netcore/sample/iOS ) from `dotnet/runtime` to highlight the concept of building applications on top of the mono runtime. 
A recent version of XCode as well as an iOS device or simulator is required to run the sample with `dotnet publish`.

The sample will depend on a NuGet Package containing the files that build the app and embed the mono runtime into the app, and the code that will build that package can be found here https://github.com/dotnet/runtime/pull/43048

### CODE REVIEW

File: core/mono-samples/iOS/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+languages:
+- c#
+products:
+- dotnet
+- iOS
+- mono runtime
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:7]: Right now this needs to be _dotnet-core_ so that it appears in the filter list for the samples browser. There isn't a product filter for _ios_ and _mono runtime_ unfortunately. We'll need to do something about this in the future though: https://review.docs.microsoft.com/en-us/samples/browse/?branch=master&expanded=dotnet&products=dotnet-core

```suggestion
products:
- dotnet-core
```

File: core/mono-samples/iOS/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+---
+languages:
+- c#
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:3]: This should be _csharp_
```suggestion
- csharp
```

File: core/mono-samples/iOS/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+NOTE: The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime.
+
+## Requirements
+
+To run the sample for iOS you will need a recent version of XCode installed (e.g. 11.3 or higher).
+
+## Building the sample
+
+To build the sample from the command line, run `dotnet publish`.
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:25]: You may want to add more details here about building. For example, see this other readme for a sample that talks about the download button on the published page:

https://review.docs.microsoft.com/en-us/samples/dotnet/samples/order-unit-tests-cs/?branch=master

File: core/mono-samples/iOS/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+# iOS Sample: Simple greeting and counter (C#)
+
+In this sample, the mono runtime is used to invoke objective-c unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
+
+NOTE: The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime.
+
+## Requirements
+
+To run the sample for iOS you will need a recent version of XCode installed (e.g. 11.3 or higher).
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:21]: This could be enhanced. Which versions of iOS? What version of Mono? Is there a link to XCode you could add? etc.

**Review by adegeo — CHANGES_REQUESTED** [review]
Thanks for adding this example. There are some changes you'll need to make so that this sample shows in the samples browser.

File: core/mono-samples/iOS/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
+languages:
+- c#
+products:
+- dotnet
+- iOS
+- mono runtime
+page_type: sample
+name: "iOS Sample: Simple greeting and counter (C#)"
+description: "An iOS application that contains an example of embedding the mono runtime to invoke unmanaged code with C#."
```
[the remark this answers is not in the document]
Comment (mdh1418) [reply core/mono-samples/iOS/README.md:8]: Will the url be generated automatically from adding this line or are there other actionables I need to take to get the sample hosted at the url?

File: core/mono-samples/iOS/README.md (hunk 1 above)
Comment (mdh1418) [reply core/mono-samples/iOS/README.md]: Will there be dotnet5 or dotnet6 added to the product filters in the future?

File: core/mono-samples/iOS/README.md (hunk 2 above)
Comment (mdh1418) [reply core/mono-samples/iOS/README.md]: Thanks!

File: core/mono-samples/iOS/README.md (hunk 3 above)
Comment (mdh1418) [reply core/mono-samples/iOS/README.md]: Added more details to outline the process from downloading the files, setting up the simulator and modifying the `.csproj`, to running the sample. Let me know how it looks!

File: core/mono-samples/iOS/README.md (hunk 4 above)
Comment (mdh1418) [reply core/mono-samples/iOS/README.md]: I received some clarifications on the minimum versions necessary and added a link to downloading Xcode.

File: core/mono-samples/iOS/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (adegeo) [reply core/mono-samples/iOS/README.md:8]: IFAIK you just need to add this line. I have heard rumors though that the samples browser doesn't automatically import samples and you need to file a ticket. I've heard it's "supposed" to be automated but it never shows up and when you ask about it, it magically shows up, as if someone is somewhere pushing a button every time.

@IEvangelist true?

File: core/mono-samples/iOS/README.md (hunk 1 above)
Comment (adegeo) [reply core/mono-samples/iOS/README.md]: No idea. I don't have anything to do with the samples. I'll bring this up to our team and see what they know.

**Comment by adegeo** [comment]
https://github.com/dotnet/samples/pull/3939 needs to be merged to fix the build error.

File: core/mono-samples/iOS/NuGet.config [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <config>
+    <add key="repositoryPath" value="/Users/mdhwang/core/mono-samples/iOS/" />
+  </config>
+  <packageSources>
+    <add key="dotnet5" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json" />
```
Comment (adegeo) [inline core/mono-samples/iOS/NuGet.config:7]: Is this URL OK? Not familiar with it..

File: core/mono-samples/iOS/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+In this sample, the mono runtime is used to invoke objective-c unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
+
+NOTE: The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
+
+## Sample Prerequisites
+
+To run the sample for iOS, Xcode and an iOS simulator is needed. 
+- Xcode: Any version should work with this sample (download Xcode at https://developer.apple.com/xcode/)
+- iOS simulator: Minimal version iOS 8.0
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:22]:
```suggestion
- iOS simulator 8.0 or greater.
```

File: core/mono-samples/iOS/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+# iOS Sample: Simple greeting and counter (C#)
+
+In this sample, the mono runtime is used to invoke objective-c unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
+
+NOTE: The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
+
+## Sample Prerequisites
+
+To run the sample for iOS, Xcode and an iOS simulator is needed. 
+- Xcode: Any version should work with this sample (download Xcode at https://developer.apple.com/xcode/)
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:21]:
```suggestion
- Xcode: Any version should work with this sample (download Xcode at https://developer.apple.com/xcode/).
```

File: core/mono-samples/iOS/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+name: "iOS Sample: Simple greeting and counter (C#)"
+description: "An iOS application that contains an example of embedding the mono runtime to invoke unmanaged code with C#."
+urlFragment: "mono-ios-csharp"
+---
+
+# iOS Sample: Simple greeting and counter (C#)
+
+In this sample, the mono runtime is used to invoke objective-c unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
+
+NOTE: The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:16]:
```suggestion
> [!NOTE]
> The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
```

File: core/mono-samples/iOS/README.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+## Sample Prerequisites
+
+To run the sample for iOS, Xcode and an iOS simulator is needed. 
+- Xcode: Any version should work with this sample (download Xcode at https://developer.apple.com/xcode/)
+- iOS simulator: Minimal version iOS 8.0
+
+NOTE: Modify `IosSimulator` under target `BuildAppBundle` from `iPhone 11` to your simulator's device name.
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:24]:
```suggestion
> [!NOTE]
> Modify `IosSimulator` under target `BuildAppBundle` from `iPhone 11` to your simulator's device name.
```

**Review by adegeo — APPROVED** [review]
I think this is looking good now. Some minor things to adjust.

File: core/mono-samples/iOS/NuGet.config [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+    <add key="mitchnuget" value="/Users/mdhwang/runtime_wasm_clean/artifacts/packages/Debug/NonShipping/" />
```
Comment (mdh1418) [inline core/mono-samples/iOS/NuGet.config:8]: This was used to test the nuget package locally. Will be modified/removed when this goes in https://github.com/dotnet/runtime/pull/43048

File: core/mono-samples/iOS/NuGet.config (hunk 6 above)
Comment (mdh1418) [reply core/mono-samples/iOS/NuGet.config:7]: I believe it contains the mono runtime? @steveisok

File: core/mono-samples/iOS/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (IEvangelist) [reply core/mono-samples/iOS/README.md:8]: My experience is that it takes a long time, roughly a week or so - but it does eventually make its way there. There have been a few times too where it has gotten stuck with no reason why, then you have to ask the right people who have access to get it moved.

File: core/mono-samples/iOS/NuGet.config [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <config>
+    <add key="repositoryPath" value="/Users/mdhwang/core/mono-samples/iOS/" />
```
Comment (akoeplinger) [inline core/mono-samples/iOS/NuGet.config:4]: this needs to be removed before merging

File: core/mono-samples/README.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# Sample .NET mono runtime
```
Comment (akoeplinger) [inline core/mono-samples/README.md:1]: I think I'd reword the header:

```suggestion
# Sample .NET apps using the Mono runtime
```

File: core/mono-samples/iOS/Program.csproj [hunk 14]
```diff
@@ -0,0 +1,55 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <RuntimeIdentifier>ios-x64</RuntimeIdentifier>
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Runtime.iOS.Sample.Mono" Version="6.0.0-dev" />
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:9]: I think this won't work with the final nuget package since it won't have a `6.0.0-dev` version like your local build but a real version like `6.0.0-alpha.1.20520.2`.

So what we can do is either hardcode a version or use `Version="6.0.0-*`.

File: core/mono-samples/iOS/Program.csproj [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  </ItemGroup>
+
+  <UsingTask TaskName="AppleAppBuilderTask"
+             AssemblyFile="$(NuGetPackageRoot)microsoft.net.runtime.ios.sample.mono/6.0.0-dev/tools/net6.0/AppleAppBuilder.dll" />
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:13]: Instead of using `NuGetPackageRoot` we should set `GeneratePathProperty="true"` on the `<PackageReference>`, that will set a property called `$(PkgMicrosoft_NET_Runtime_iOS_Sample_Mono)` which points to the directory:

```suggestion
             AssemblyFile="$(PkgMicrosoft_NET_Runtime_iOS_Sample_Mono)\tools\net6.0\AppleAppBuilder.dll" />
```

File: core/mono-samples/iOS/Program.csproj [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+  <Target Name="BuildAppBundle" AfterTargets="CopyFilesToPublishDirectory;Publish;BuildAppleAppBuilder">
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:15]: I don't see the `BuildAppleAppBuilder`  target anywhere, is it needed?
Additionally, I think we only need to run after the Publish target, not CopyFilesToPublishDirectory.

File: core/mono-samples/iOS/Program.csproj [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    </PropertyGroup>
+
+    <ItemGroup>
+      <BundleAssemblies Include="$(AppDir)\*.dll" />
+    </ItemGroup>
+
+    <AppleAppBuilderTask
+      ProjectName="HelloiOS"
+      AppDir="$(AppDir)"
+      MonoRuntimeHeaders="%(ResolvedRuntimePack.PackageDirectory)\runtimes\ios-x64\native\include\mono-2.0"
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:30]:
```suggestion
      MonoRuntimeHeaders="%(ResolvedRuntimePack.PackageDirectory)\runtimes\$(RuntimeIdentifier)\native\include\mono-2.0"
```

File: core/mono-samples/iOS/README.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+- dotnet-core
+page_type: sample
+name: "iOS Sample: Simple greeting and counter (C#)"
+description: "An iOS application that contains an example of embedding the mono runtime to invoke unmanaged code with C#."
+urlFragment: "mono-ios-csharp"
+---
+
+# iOS Sample: Simple greeting and counter (C#)
+
+In this sample, the mono runtime is used to invoke objective-c unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
```
Comment (akoeplinger) [inline core/mono-samples/iOS/README.md:14]:
```suggestion
In this sample, the mono runtime is used to invoke Objective-C unmanaged code (main.m) from the C# managed side (Program.cs) and vice versa. With the sample running, you can enter your name and click the corresponding button to modify the greeting message as well as clicking a button to increment a counter.
```

File: core/mono-samples/iOS/README.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+[!NOTE]
+The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
+
+## Sample Prerequisites
+
+To run the sample for iOS, Xcode and an iOS simulator is needed.
```
Comment (akoeplinger) [inline core/mono-samples/iOS/README.md:21]: probably good idea to mention that this only works on macOS.

File: core/mono-samples/iOS/Program.csproj [hunk 20]
```diff
@@ -0,0 +1,55 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:3]: should this be `net6.0`? it probably should, but it's possible the dotnet sdk doesn't support it yet.

we might want to mention in the readme that we require a daily master build of the dotnet sdk from https://github.com/dotnet/installer.

**Review by akoeplinger — COMMENTED** [review]
Left a few comments, looks good overall :)

File: core/mono-samples/iOS/Program.csproj (hunk 16 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj]: Ah right, removing both.

File: core/mono-samples/iOS/Program.csproj (hunk 15 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj]: Thanks!

File: core/mono-samples/iOS/NuGet.config (hunk 12 above)
Comment (mdh1418) [reply core/mono-samples/iOS/NuGet.config]: Done

File: core/mono-samples/iOS/Program.csproj (hunk 20 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj:3]: I tried `net6.0` a while ago, and I believe the dotnet sdk did not support it at that time.
Will add the note about the dotnet sdk build.

File: core/mono-samples/iOS/Program.csproj (hunk 14 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj]: Thanks!

File: core/mono-samples/iOS/Program.csproj [hunk 21]
```diff
@@ -6,13 +6,13 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Runtime.iOS.Sample.Mono" Version="6.0.0-dev" />
+    <PackageReference Include="Microsoft.NET.Runtime.iOS.Sample.Mono" Version="6.0.0-*" GeneratePathProperty="true"/>
```
Comment (akoeplinger) [inline core/mono-samples/iOS/Program.csproj:9]: Looks like this doesn't work on CI because this repo installs the 5.0 dotnet sdk. I tried locally and what works is using `$(BundledNETCorePlatformsPackageVersion)` as the version since then it'll try to get the same version as it uses for the runtime pack.

That brings up another issue though, right now it tries to restore `Microsoft.NETCore.App.Runtime.ios-x64 with version (= 5.0.0)` but the final 5.0 release isn't published to nuget.org yet. And even if it were we don't really want to use the 5.0 version.

As far as I know the only way to consume 6.0 assets is by installing nightly master builds of the dotnet sdk from the dotnet/installer repo. I don't know if we want to do that in CI in this repo (maybe we just need to skip this project there).

Additionally we can add the following global.json next to Program.csproj to force usage of the 6.0 SDK:

```
{
  "sdk": {
    "version": "6.0.100-alpha",
    "rollForward": "major",
    "allowPrerelease": true
  }
}
```

Comment (steveisok) [reply core/mono-samples/iOS/Program.csproj:9]: Yeah, I think skipping the project makes sense for now.  It's in a bit of limbo.

Comment (steveisok) [reply core/mono-samples/iOS/Program.csproj:9]: @adegeo Tagging you because you made the last commit to the yml.  Can we add another leg that installs the net6 preview sdk and builds the iOS project?  That seems to be the only way to be correct on CI for us. Otherwise, we can just skip the project.

Comment (adegeo) [reply core/mono-samples/iOS/Program.csproj:9]: @steveisok Sorry, missed this alert. Yeah we could do that easily as long as the dotnet-install powershell script supports it?

Comment (steveisok) [reply core/mono-samples/iOS/Program.csproj:9]: @adegeo Looks like it does with the `--version` switch.  Would it be easier to bump the sdk version in CI or just add another leg?

Comment (adegeo) [reply core/mono-samples/iOS/Program.csproj:9]: @steveisok If you can give me the exactly command line I'll edit the script to add a leg.

**Comment by adegeo** [comment]
Depends on #3941

File: core/mono-samples/iOS/README.md [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+[!NOTE]
+The purpose of this sample is to demonstrate the concept of building an iOS application on top of the mono runtime. The mono runtime headers should be supplied through the build process.
+
+## Sample Prerequisites
+
+This sample will only run on macOS as it requires Xcode and an iOS simulator.
+
+- Xcode: Any version should work with this sample (download Xcode at <https://developer.apple.com/xcode/>).
+- iOS simulator 8.0 or greater.
+- Daily master build of the dotnet sdk (<https://github.com/dotnet/installer>).
```
Comment (adegeo) [inline core/mono-samples/iOS/README.md:25]: @steveisok @mdh1418 @akoeplinger 
Considering it's not easy to install the .NET 6 SDK at the moment, I would improve this with the specific commands to install. I don't see anyone figuring out how to install. I couldn't and had to have @steveisok help.

**Review by adegeo — CHANGES_REQUESTED** [review]
Once we address the comment I made about installing .NET 6, I think we can merge this.

File: core/mono-samples/iOS/README.md (hunk 22 above)
Comment (mdh1418) [reply core/mono-samples/iOS/README.md]: Added more details to install the specific dotnet version. Was this what you meant?

File: core/mono-samples/iOS/Program.csproj (hunk 21 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj:9]: Fixed the version at `6.0.100-alpha.1.20531.2`.

File: core/mono-samples/iOS/Program.csproj (hunk 20 above)
Comment (mdh1418) [reply core/mono-samples/iOS/Program.csproj]: Updated it to be `net6.0`.

File: core/mono-samples/README.md [hunk 23]
```diff
@@ -0,0 +1,7 @@
+# Sample .NET apps using the Mono runtime
+
+This folder contains sample code demonstrating how to build mobile (iOS, Android, Browser WebAssembly) apps with the mono runtime. The requirements for each sample can be found within the corresponding folder, and each sample can be ran using `dotnet publish`.
+
+## Prerequisites
+
+Dotnet sdk version 6.0.100-alpha.1.20531.2 is needed to run these samples. To install a specific version of the dotnet sdk, download the latest stable version of  `dotnet-install.sh` at <https://github.com/dotnet/install-scripts> and execute the script with `./dotnet-install.sh --version 6.0.100-alpha.1.20531.2`. Ensure that the version is correct with `dotnet --version` as the path may not be properly set.
```
Comment (adegeo) [inline core/mono-samples/README.md:7]: Let's make this a bit more generic. I use the powershell script for example. You'll have to merge this by hand as I didn't use full code blocks because the suggestion feature here won't work with it. Meaning, I used two backticks for the code blocks (bash and powershell) instead of three.

I also don't understand this clause:  Ensure that the version is correct with `dotnet --version` as the path may not be properly set.

```suggestion
.NET SDK version 6.0.100-alpha.1.20531.2 is needed to run these samples. To install a specific version of the dotnet sdk, download the latest stable version of  the dotnet-install script:

- Bash (Linux/macOS): <https://dot.net/v1/dotnet-install.sh>
- PowerShell (Windows): <https://dot.net/v1/dotnet-install.ps1>

Install version .NET version **6.0.100-alpha.1.20531.2**:

``bash
./dotnet-install.sh --version 6.0.100-alpha.1.20531.2
``
``powershell
./dotnet-install.ps1 -Version 6.0.100-alpha.1.20531.2
``
For more information, see [dotnet-install script reference](https://docs.microsoft.com/dotnet/core/tools/dotnet-install-script)
```

Comment (mdh1418) [reply core/mono-samples/README.md:7]: Thanks! 
Ah, for more context, when installing dotnet sdk from one of the links in the chart at https://github.com/dotnet/installer, I believe the sdk will be installed at `/usr/local/share/dotnet/dotnet`. 
When I ran the `dotnet-install.sh` script to install `version 6.0.100-alpha.1.20531.2`, it was neither installed there nor at the default value of `/usr/share/dotnet` (`--InstallDir` under https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script)
I only realized this when checking the version of dotnet and seeing that it didn't match `6.0.100-alpha.1.20531.2` because my path pointed to `/usr/local/share/dotnet/dotnet`.
Instead, I saw the message: `dotnet-install: Adding to current process PATH: `/Users/mdhwang/.dotnet`. Note: This change will be visible only when sourcing script.` so it would be helpful for the user to check which dotnet will be used by checking the version and adding the correct path.
Note: I tried sourcing the script but it kept crashing the terminal window

Comment (mdh1418) [reply core/mono-samples/README.md:7]: I suppose that the users can find out about `-InstallDir` from going to the docs as well

**Comment by mdh1418** [comment]
@adegeo I made a few more changes and think this PR is ready for another look over.

**Review by adegeo — APPROVED** [review]
